tests/sentry/api/test_event_search.py

```diff
@@ -768,74 +768,107 @@ def test_empty_string(self):
-        self.hi = ["equals", ["user.email", "hi@example.com"]]
+        users = ["foo", "bar", "foobar", "hello", "hi"]
+        for u in users:
+            self.__setattr__(u, ["equals", ["user.email", "{}@example.com".format(u)]])
+            self.__setattr__("o{}".format(u), ["user.email", "=", "{}@example.com".format(u)])
 
     def test_simple(self):
         result = get_filter("user.email:foo@example.com OR user.email:bar@example.com")
-        assert result.conditions == [[["or", [self.foo, self.bar]], "=", 1]]
+        assert result.conditions == [[_or(self.foo, self.bar), "=", 1]]
 
         result = get_filter("user.email:foo@example.com AND user.email:bar@example.com")
-        assert result.conditions == [[["and", [self.foo, self.bar]], "=", 1]]
+        assert result.conditions == [self.ofoo, self.obar]
 
     def test_single_term(self):
         result = get_filter("user.email:foo@example.com")
-        assert result.conditions == [["user.email", "=", "foo@example.com"]]
+        assert result.conditions == [self.ofoo]
 
     def test_order_of_operations(self):
         result = get_filter(
             "user.email:foo@example.com OR user.email:bar@example.com AND user.email:foobar@example.com"
         )
-        assert result.conditions == [[["or", [self.foo, ["and", [self.bar, self.foobar]]]], "=", 1]]
+        assert result.conditions == [[_or(self.foo, _and(self.bar, self.foobar)), "=", 1]]
 
         result = get_filter(
             "user.email:foo@example.com AND user.email:bar@example.com OR user.email:foobar@example.com"
         )
-        assert result.conditions == [[["or", [["and", [self.foo, self.bar]], self.foobar]], "=", 1]]
+        assert result.conditions == [[_or(_and(self.foo, self.bar), self.foobar), "=", 1]]
 
     def test_multiple_statements(self):
         result = get_filter(
             "user.email:foo@example.com OR user.email:bar@example.com OR user.email:foobar@example.com"
         )
-        assert result.conditions == [[["or", [self.foo, ["or", [self.bar, self.foobar]]]], "=", 1]]
+        assert result.conditions == [[_or(self.foo, _or(self.bar, self.foobar)), "=", 1]]
 
         result = get_filter(
             "user.email:foo@example.com AND user.email:bar@example.com AND user.email:foobar@example.com"
         )
-        assert result.conditions == [
-            [["and", [self.foo, ["and", [self.bar, self.foobar]]]], "=", 1]
-        ]
+        assert result.conditions == [self.ofoo, self.obar, self.ofoobar]
 
         # longer even number of terms
         result = get_filter(
             "user.email:foo@example.com AND user.email:bar@example.com OR user.email:foobar@example.com AND user.email:hello@example.com"
         )
         assert result.conditions == [
-            [["or", [["and", [self.foo, self.bar]], ["and", [self.foobar, self.hello]]]], "=", 1]
+            [_or(_and(self.foo, self.bar), _and(self.foobar, self.hello)), "=", 1]
```

Comment:
Would we be able to optimize sibling AND conditions inside an OR later on?

Reply:
Unfortunately not. These optimizations are bound by the current json schema. There is a small potential for collapsing OR trees when they are inside AND conditions, since the json schema allows for that. I looked at it, and I think it would be making the code even harder to understand without any real benefit though.